composer/callbacks/activation_monitor.py

```diff
@@ -0,0 +1,216 @@
+        self.handles.append(module.register_forward_hook(partial(self.forward_hook, logger, step)))
+
+    def forward_hook(self, logger: Logger, step: Optional[int], module: torch.nn.Module, input: Sequence,
+                     output: Sequence):
+        module_name = self.module_names[module]
+
+        if self.ignore_module_types is not None:
+            for ignore_module_type in self.ignore_module_types:
+                if ignore_module_type in module_name:
+                    return
+
+        metrics = {}
+
+        for i, val in enumerate(input):
+            if val is None or isinstance(val, dict):
+                continue
+            self.recursively_add_metrics(metrics, module_name, f'_input.{i}', val)
+
+        for i, val in enumerate(output):
+            if val is None or isinstance(val, dict):
+                continue
+            self.recursively_add_metrics(metrics, module_name, f'_output.{i}', val)
+
+        if self.only_log_wandb:
+            wandb_loggers = [ld for ld in logger.destinations if isinstance(ld, WandBLogger)]
+            if len(wandb_loggers):
+                for wandb_logger in wandb_loggers:
+                    wandb_logger.log_metrics(metrics, step)
+            else:
+                # In the case there were no WandB loggers, just default to
+                # the standard logger and let it take care of it
+                logger.log_metrics(metrics)
+        else:
+            logger.log_metrics(metrics)
+
+    def recursively_add_metrics(self, metrics: dict, name: str, suffix: str, values: Any):
+        # Keep recursively diving if the value is a sequence
+        if isinstance(values, Sequence):
+            for i, value in enumerate(values):
+                self.recursively_add_metrics(metrics, f'{name}_{i}', suffix, value)
+            return
+        else:
+            self.add_metrics(metrics, name, suffix, values)
+
+    def add_metrics(self, metrics: dict, name: str, suffix: str, value: torch.Tensor):
+        # We shouldn't log booleans
+        if value.dtype == torch.bool:
+            return
+        if value.is_floating_point() or value.is_complex():
+            metrics[f'activations/l2_norm/{name}{suffix}'] = torch.linalg.vector_norm(value, dim=-1).mean().item()
+            metrics[f'activations/average/{name}{suffix}'] = value.mean().item()
+            metrics[f'activations/kurtosis/{name}{suffix}'] = self.compute_kurtosis(value).item()
+
+            # Because we call max with `dim=-1` we need to call .values to get the actual values
+            metrics[f'activations/max/{name}{suffix}'] = value.max(dim=-1).values.mean().item()
+
+    def compute_kurtosis(self, value: torch.Tensor):
+        # Computes the kurtosis over the last dimension
+        mean = torch.mean(value, dim=-1).unsqueeze(-1)
+        diffs = value - mean
+        m_4 = torch.mean(torch.pow(diffs, 4), dim=-1)
+        var = torch.mean(torch.pow(diffs, 2), dim=-1)
+        return (m_4 / (var**2)).mean()
```

Comment:
I think you _may_ not need this last `mean` call...


It may be nice to either test these calculations against this https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.kurtosis.html

Or just use scipy directly.

Reply:
I think because of how we're computing statistics, for a batch of size (batch_size, seq_len, d_model) basically the `(m_4 / (var**2))` ends up being (batch_size, seq_len) so we need the reduction.